### PR TITLE
Added additional authentication levels AUTH_EXPOSE, AUTH_HISTORY, AUT…

### PR DESCRIPTION
**Pull request by ahmetsacan**
This is a follow up to (https://github.com/dokuwiki/dokuwiki/pull/4303). Please see the discussion there.

### CODE REVIEW

**Review by fiwswe — COMMENTED** [review]
General comments:
* Tests are still missing.
* Even though using the float values to insert new permission types seems ok at first glance, there may still be hidden consequences. You'd need to prove that this does not cause any problems in core, plugins or templates.
* If (!) this PR is approved then documentation on [dokuwiki.org](https://dokuwiki.org) would also need to be updated.
* As @splitbrain mentioned in #4303 there may be ways to achieve your goals without touching DokuWiki core.
  * Some ideas off the top of my head for example:
    * Idea 1
      1. Define a new group.
      1. Then add your students to this group.
      1. Create a plugin that parses some syntax on the page to trigger the plugin and to read a group name as a parameter. Then check if the current user is a member of the group and trigger functionality accordingly. The [hidepages Plugin](https://www.dokuwiki.org/plugin:hidepages) may be a starting point, even though it is unmaintained and probably no longer functioning with Kaos.
      1. Add the trigger syntax to the pages you want to treat special.
    * Idea 2
      * Write a plugin that treats pages in certain namespaces in a special way.
    * Idea 3
      * Write a plugin that treats pages whose page ID matches a certain pattern, e.g. a prefix or suffix, in a special way.
    * For the use case of the link that allows access to an otherwise hidden page @splitbrain had a rough suggestion in the other PR.

File: inc/Action/Diff.php [hunk 1]
```diff
@@ -17,7 +17,7 @@ class Diff extends AbstractAction
     /** @inheritdoc */
```
Comment (fiwswe) [inline inc/Action/Diff.php:17]: https://github.com/dokuwiki/dokuwiki/blob/df992f97cb02d677b035f6035cb1515b2ae3a3d1/inc/Action/AbstractAction.php#L42

Needs to be updated as well.

**Comment by splitbrain** [comment]
@fiwswe already summed it up pretty well. I currently see no reason why this should be applied. All stated goals can be achieved with plugins already.

File: inc/Extension/PluginController.php [hunk 2]
```diff
@@ -107,7 +107,7 @@ public function load($type, $name, $new = false, $disabled = false)
             if (!class_exists($class, true)) {
                 # the plugin might be in the wrong directory
                 $inf = confToHash(DOKU_PLUGIN . "$plugin/plugin.info.txt");
-                if ($inf['base'] && $inf['base'] != $plugin) {
+                if (isset($inf['base']) && $inf['base'] && $inf['base'] != $plugin) {
```
[the remark this answers is not in the document]
Comment (ahmetsacan) [reply inc/Extension/PluginController.php:110]: Yes, unrelated to the current issue. The unit tests were showing php warnings for undefined array key here, which is why I added isset()

File: inc/auth.php [hunk 3]
```diff
@@ -799,7 +799,7 @@ function auth_aclcheck_cb($data)
             }
             //we had a match - return it
             if ($perm != -1) {
-                return (int) $perm;
+                return (float) $perm;
```
[the remark this answers is not in the document]
Comment (ahmetsacan) [reply inc/auth.php:802]: updated.

File: inc/auth.php [hunk 4]
```diff
@@ -769,7 +769,7 @@ function auth_aclcheck_cb($data)
         }
         if ($perm > -1) {
             //we had a match - return it
-            return (int) $perm;
+            return (float) $perm;
```
[the remark this answers is not in the document]
Comment (ahmetsacan) [reply inc/auth.php:772]: updated.

File: inc/Action/Diff.php (hunk 1 above)
Comment (ahmetsacan) [reply inc/Action/Diff.php:17]: updated.

File: inc/auth.php (hunk 4 above)
[the remark this answers is not in the document]
Comment (fiwswe) [reply inc/auth.php:772]: No commit for the update is visible here on GitHub.

File: inc/Extension/PluginController.php (hunk 2 above)
[the remark this answers is not in the document]
Comment (fiwswe) [reply inc/Extension/PluginController.php:110]: Then that should be separated into its own PR. For this one the original code should stay as it was.

**Review by fiwswe — COMMENTED** [review]
Mostly irrelevant because this PR will not be merged. Just giving feedback to your reply.

**Comment by ahmetsacan** [comment]
> * Tests are still missing.

I updated test cases to _test/tests/Action/ActionTest.php _test/tests/inc/auth_aclcheck.test.php


@fiwswe and @splitbrain, Thank you for the ideas for implementing this as a plugin. I don't think a plugin will allow me to cleanly/securely/conveniently implement what I am trying to accomplish. The most appropriate action hook seems to be AUTH_ACL_CHECK triggered in auth_aclcheck(), but the data does not contain what I need; specifically it is missing the "action" that the ACL level is being checked for. The discrepancy between actions and acl permission seem to be the source of my difficulties. To implement this cleanly as a plugin, consolidating/refactoring the auth ACL checks into a function like @fiwswe suggested would be great, but I would suggest distinguishing between $requestedPermission and $requestedAction (the purpose for checking the permission). The core may ignore the requestedAction in this function, but propagating it in the event data would allow plugins to make their own interpretation:

```php
function auth_HasSufficientPermissionsFor(int $requestedPermission, ?string $id, ?string $user=null, array $groups=[], ?string $requestedAction): bool
{
    $auth = auth_aclcheck($id, $user, $groups);
    return $requestedPermission <= $auth;  // Simulate the current functionality.
}
```

If such a central function is not enticing, you may also consider triggering an event in the Action classes, e.g., in the minimumPermission() functions, to allow plugins to customize the minimumPermission required for an action. I personally wouldn't be a fan of using this approach for my needs, because there are requests in core and other plugins that do not use the Action framework when deciding on permissions.


Aside from whether this can be accomplished in a plugin, there is still also the issue of how the configuration (the settings for who can expose/source/history for which pages) can be stored in a customizable manner. I just don't see the point in a plugin replicating its own ACL configuration to store these settings for pages/users/groups. 

I believe we are at an impasse. You are not convinced of the need for these refined permission levels and we disagree whether they can cleanly and conveniently be implemented in a plugin. I will use my proposed changes for my own site, but I will stop sending pull requests for this.

Thank you for your consideration and feedback.

**Comment by splitbrain** [comment]
The current action is accessible through a global.  There's also ACTION_ACT_PREPROCESS which could easily check users and actions. I recommend looking at the source code of other plugins ding similar things.

**Comment by fiwswe** [comment]
>Aside from whether this can be accomplished in a plugin, there is still also the issue of how the configuration (the settings for who can expose/source/history for which pages) can be stored in a customizable manner. I just don't see the point in a plugin replicating its own ACL configuration to store these settings for pages/users/groups.

You are too fixated on solving this using modified ACLs IMHO. Like I wrote there are other methods to accomplish the same goals. And they would not rely on `AUTH_ACL_CHECK`. I'd recommend stepping back from your current solution and thinking _out of the box_ to find solutions that can be implemented by plugins.

**Comment by fiwswe** [comment]
>… consolidating/refactoring the auth ACL checks into a function like @fiwswe suggested would be great …

This would only make sense if a general overhaul of the ACL/permission system was planned for DokuWiki. And it would be more of a long term thing, as it would take time for plugins & templates to abandon the current code in favour of the new API. It would probably require deprecating certain functions for one or more major releases before pulling the plug.

It would also need more thoughts and discussions, not the quick-and-dirty idea I suggested.

From what @splitbrain wrote that is rather unlikely. And maybe its not even needed.

**Comment by fiwswe** [comment]
> > * Tests are still missing.
> 
> I updated test cases to _test/tests/Action/ActionTest.php _test/tests/inc/auth_aclcheck.test.php

Maybe you did in your local repository. But the changes did not make it here. Check the »Files changed« tab of this PR.

**Comment by ahmetsacan** [comment]
> Maybe you did in your local repository. But the changes did not make it here. Check the »Files changed« tab of this PR.

Maybe changes in the newer commits don't show up once the PR is closed?

The changes can be seen in my repo:
https://github.com/dokuwiki/dokuwiki/compare/master...ahmetsacan:dokuwiki:master

Note that I am now treating my fork as a sandbox, so you may see changes there that are unrelated to this PR in the future.

**Comment by fiwswe** [comment]
>Maybe changes in the newer commits don't show up once the PR is closed?

Could be the reason.

>Note that I am now treating my fork as a sandbox, so you may see changes there that are unrelated to this PR in the future.

That's why it's a good idea to use dedicated branches for PRs. ;-)

**Comment by ahmetsacan** [comment]
> That's why it's a good idea to use dedicated branches for PRs. ;-)

Makes sense.  I am new to PR, thanks for your understanding.

**Pull request closed without merging**